WeakAuras/Transmission.lua

```diff
@@ -316,37 +270,318 @@ ChatFrame_AddMessageEventFilter("CHAT_MSG_BN_WHISPER_INFORM", filterFunc)
+    anchorFrameFrame = checkButtons[8],
+    yOffset          = checkButtons[8],
+    -- not super convinced that these are trivial
+    height           = checkButtons[8], 
+    width            = checkButtons[8],
+    fontSize         = checkButtons[8], 
+    -- trigger
+    trigger             = checkButtons[6],
+    untrigger           = checkButtons[6],
+    disjunctive         = checkButtons[6],
+    additional_triggers = checkButtons[6],
+    activeTriggerMode   = checkButtons[6],
+    numTriggers         = checkButtons[6],
+    -- conditions
+    conditions = checkButtons[5],
+    -- load
+    load = checkButtons[4],
+    -- actions
+    actions = checkButtons[3],
+    -- animation
+    animation = checkButtons[2],
+  },
+  {
+    __index = function(self, key) return key and checkButtons[7] end
+  }
+)
+
+buttonAnchor.import:SetScript("OnClick", function(self)
+    if not self.data then return end
+    -- build closure for DynFrame to use
+    HideUIPanel(ItemRefTooltip)
+    WeakAuras.LoadOptions()
+    WeakAuras.SetImporting(true)
+    buttonAnchor.showCode.codes = nil
+    buttonAnchor.thumbnail.currentThumbnail:Hide()
+    buttonAnchor.thumbnail.currentThumbnail = nil
+    -- TODO: find a way to do this better because uh
+    -- it's ugly as hell
+    local optionsFrame = WeakAuras.OptionsFrame()
+    if not(optionsFrame) then
+      WeakAuras.ToggleOptions()
+      optionsFrame = WeakAuras.OptionsFrame()
+    end
+    if not(WeakAuras.IsOptionsOpen()) then
+      WeakAuras.ToggleOptions()
+    end
+    -- TODO: this function is used in at least one other file and needs to be de-duplicated
+    local function findUnusedId(id)
+      local num = 2;
+      while WeakAurasSaved.displays[id] do
+        id = id.." "..num;
+        num = num + 1;
+      end
+      return id
+    end
+    -- upvalues for our closure, and also clear data so that the next import is not corrupted
+    local incomingData, incomingChildren, diffs = self.data, self.children, self.diffs
+    self.data, self.children, self.diffs = nil, nil, nil
+    local mode = (not diffs and "import") or (checkButtons[1]:GetChecked() and "copy") or "update"
+    WeakAuras.dynFrame:AddAction("import", coroutine.create(
+        function()
+          incomingData.parent = nil
+          if mode ~= "update" or diffs[0] == "added" then
+            if mode == "copy" then 
+              incomingData.uid = nil
+            end
+            incomingData.id = findUnusedId(incomingData.id)
+            WeakAuras.Add(incomingData)
+          else
+            -- clear fields which the user did not want to update
+            local oldData, patch = unpack(diffs[0])
+            if patch then
+              for k in pairs(patch) do
+                local checkbutton = keyToButton[k]
+                if checkbutton and not checkbutton:GetChecked() then
+                  patch[k] = nil
+                end
+              end
+              if patch.id then -- ensure that the id is changed
```

Comment:
TODO: I don't understand this.

Reply:
Are you talking about where I also assign the unused id to `incomingData`, even though i'm not going to use the `incomingData` to actually import anything? This is because `patch` is too small in scope to reach line 498, where I need the ID of the installed/updated/copied aura so that I can call `GetData` properly - this allows me to only write `GetData` once, instead of copying it into each branch of the if-else chain. As for why I need to call GetData at all, this was a workaround to a pretty nasty bug which caused the parent of an updated aura to think it did not have the updated aura as a child anymore.